swayidle/main.c

```diff
@@ -0,0 +1,397 @@
+bool setup_inhibit = true;
+
+struct swayidle_state {
+	struct wl_display *display;
+	struct org_kde_kwin_idle_timeout *idle_timer;
+	struct org_kde_kwin_idle_timeout *lock_timer;
+	struct wlr_output_layout *layout;
+} state;
+
+
+static void invoke_swaylock() {
+	wlr_log(L_INFO,"Requesting locking screen");
+	uint32_t size = 4;
+	ipc_single_command(ipc_socketfd, IPC_COMMAND, "lock", &size);
+}
+
+#if defined(SWAY_IDLE_HAS_SYSTEMD) || defined(SWAY_IDLE_HAS_ELOGIND)
+
+bool have_lock() {
+	const char *dir = getenv("XDG_RUNTIME_DIR");
+	if (!dir) {
+		dir = "/tmp";
+	}
+	const char *disp = getenv("WAYLAND_DISPLAY");
+	if (!disp) {
+		disp = "default";
+	}
+	char path[256];
+	if (255 <= snprintf(path, 255,
+				"%s/swaylock.%s.%i.active", dir, disp, getuid())) {
+		wlr_log(L_INFO,"Lock path to long");
+		return false;
+	}
+	FILE *file;
+	if ((file = fopen(path, "r")))
+	{	
+		fclose(file);
+		wlr_log(L_DEBUG, "Found lockscreen");
+		return true;
+	}
+	wlr_log(L_DEBUG, "No lockscreen");
+	return false;
+}
+
+static int lock_fd = -1;
+static int ongoing_fd = -1;
+
+static int release_lock(void *data) {
+	wlr_log(L_INFO, "Releasing sleep lock %d", ongoing_fd);
+	if (ongoing_fd>=0) {
+		close(ongoing_fd);
+	}
+	ongoing_fd = -1;
+	return 0;
+}
+
+void aquire_sleep_lock() {
+	sd_bus_message *msg = NULL;
+	sd_bus_error error = SD_BUS_ERROR_NULL;
+	struct sd_bus *bus;
+	int ret = sd_bus_default_system(&bus);
+	//int ret = sd_bus_open_system(&bus);
+	if (ret < 0) {
+		wlr_log(L_ERROR, "Failed to open D-Bus connection: %s", strerror(-ret));
+		return;
+	}
+
+	ret = sd_bus_call_method(bus, "org.freedesktop.login1",
+			"/org/freedesktop/login1", "org.freedesktop.login1.Manager", "Inhibit",
+			&error, &msg, "ssss", "sleep", "sway-idle", "Setup Up Lock Screen", "delay");
+	if (ret < 0) {
+		wlr_log(L_ERROR, "Failed to send Inhibit signal: %s",
+				strerror(-ret));
+	} else {
+		ret = sd_bus_message_read(msg, "h", &lock_fd);
+		if (ret < 0) {
+			wlr_log(L_ERROR, "Failed to parse D-Bus response for Inhibit: %s", strerror(-ret));
+		}
+	}
+	wlr_log(L_INFO, "Got sleep lock: %d", lock_fd);
+}
+
+static int prepare_for_sleep(sd_bus_message *msg, void *userdata, sd_bus_error *ret_error) {
+	bool going_down = true;
+	int ret = sd_bus_message_read(msg, "b", &going_down);
+	if (ret < 0) {
+		wlr_log(L_ERROR, "Failed to parse D-Bus response for Inhibit: %s", strerror(-ret));
+	}
+	wlr_log(L_DEBUG, "PrepareForSleep signal received %d", going_down);
+	if (!going_down) {
+		aquire_sleep_lock();
+		return 0;
+	}
+
+	ongoing_fd = lock_fd;
+	if(have_lock()) {
+		wlr_log(L_DEBUG, "Have lock, no inhibit");
+		release_lock(NULL);
+		return 0;
+	}
+
+	invoke_swaylock();
+
+	if (ongoing_fd < 0) {
+		wlr_log(L_DEBUG, "Failed to get lock, we are done");
+		return 0;
+	}
+	struct timespec tv;
+	tv.tv_sec=0;
+	tv.tv_nsec = 100000000L;
+
+	wlr_log(L_DEBUG, "Checking for lock");
+	int inhibit_cnt = 0;
+	while(!have_lock() && inhibit_cnt < 5) {
+		nanosleep(&tv, NULL);
```

Comment:
Polling is evil.

Reply:
Why? This is in its own thread, so if it wasn't sleeping it would still not do anything. This is for all practical reasons completely equivalent of having a timer on a eventloop, since this thread doesn't have any other thing to do. The difference in this case would be absolutely zero.

Reply:
>Why? This is in its own thread

Oh god why does this have threads

Reply:
Because this is just a huge, ugly hack. This isn't how you "wait for an event be be triggered".

Reply:
sd_bus_wait will block waiting for the next request, so it felt easier to just have a simple thread than to start up a full eventloop just for this...

Reply:
Threads are the devil. Can't you somehow get the underlying dbus fd and poll (as in poll.h) on it?